modules/metastorage-common/src/main/java/org/apache/ignite/metastorage/common/WatchListener.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.metastorage.common;
+
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * The listener which receives and handles watch updates.
+ */
+public interface WatchListener {
+    /**
+     * The method will be called on each meta storage update.
+     *
+     * @param events A single event or batch.
+     * @return {@code True} if listener must continue event handling. If returns {@code false} then the listener and
+     * corresponding watch will be unregistered.
+     */
+    boolean onUpdate(@NotNull Iterable<WatchEvent> events);
```

Comment:
Is it guaranteed that the batch will contain only updates from a single multi-update or multiple updates can be passed to this method? We need to state this clearly in the javadoc.

Reply:
Additionally, what thread will this be executed? Are all invocations of onUpdate serialized (essentially, single-threaded?). Does one WatchListener execution affect other listeners?

Reply:
@agoncharuk 
> Is it guaranteed that the batch will contain only updates from a single multi-update or multiple updates can be passed to this method? We need to state this clearly in the javadoc.

Thanks for spotting this. Added to the `events` parameter javadoc:

> The batch always contains updates for specific revision.

Reply:
@agoncharuk 
 
> Additionally, what thread will this be executed? Are all invocations of onUpdate serialized (essentially, single-threaded?). Does one WatchListener execution affect other listeners?

It isn't defined by interface yet. What do you mean when talk that `WatchListener` execution affects other listeners?